System/Console/ANSI/Windows/Emulator.hs

```diff
@@ -339,3 +346,53 @@ toANSIColor color = fst $ minimumBy order aNSIColors
                  dg = g' - g
                  db = b' - b
              in  dr * dr + dg * dg + db * db
+
+-- getReportedCursorPosition :: IO String
+-- (See Common-Include.hs for Haddock documentation)
+getReportedCursorPosition = catch getReportedCursorPosition' getCPExceptionHandler
+  where
+    getReportedCursorPosition' = withHandleToHANDLE stdin action
+      where
+        action hdl = do
+            n <- getNumberOfConsoleInputEvents hdl
+            if n == 0
+                then return ""
+                else do
+                    es <- readConsoleInput hdl n
+                    return $ stringFromInputEvents es
+        stringFromInputEvents = cWcharsToChars . wCharsFromInputEvents
+        wCharsFromInputEvents = mapMaybe wCharFromInputEvent
+        wCharFromInputEvent e = if isKeyDownEvent
+            then Just (unicodeAsciiChar $ keyEventChar keyEventRecord)
+            else Nothing
+          where
+            eventType = inputEventType e
+            InputKeyEvent keyEventRecord = inputEvent e
+            isKeyDown = keyEventKeyDown keyEventRecord
+            isKeyDownEvent = eventType == 1 && isKeyDown
+
+-- getCursorPosition :: IO (Maybe (Int, Int))
+-- (See Common-Include.hs for Haddock documentation)
+getCursorPosition = catch getCursorPosition' getCPExceptionHandler
+  where
+    getCursorPosition' = do
+        withHandleToHANDLE stdin flush -- Flush the console input buffer
+        reportCursorPosition
+        hFlush stdout -- ensure the report cursor position code is sent to the
+                      -- operating system
+        input <- getReportedCursorPosition
+        case readP_to_S cursorPosition input of
+            [] -> return Nothing
+            [((row, col),_)] -> return $ Just (row, col)
+            (_:_) -> return Nothing
+      where
+        flush hdl = do
+            n <- getNumberOfConsoleInputEvents hdl
+            unless (n == 0) (void $ readConsoleInput hdl n)
+
+getCPExceptionHandler :: SomeException -> IO a
```

Comment:
Since you have a particular exception in mind here, it's better to restrict the type of the handler to that exception alone, to avoid false positives.

Reply:
`IOException` now replaces `SomeException`.